codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,356 @@
+		descriptor.FieldDescriptorProto_TYPE_INT64:  true,
+		descriptor.FieldDescriptorProto_TYPE_UINT32: true,
+		descriptor.FieldDescriptorProto_TYPE_UINT64: true,
+		descriptor.FieldDescriptorProto_TYPE_SINT32: true,
+		descriptor.FieldDescriptorProto_TYPE_SINT64: true,
+		descriptor.FieldDescriptorProto_TYPE_BOOL:   true,
+		descriptor.FieldDescriptorProto_TYPE_ENUM:   true,
+	},
+
+	// "1	64-bit:	fixed64, sfixed64, double"
+	1: {
+		descriptor.FieldDescriptorProto_TYPE_FIXED64:  true,
+		descriptor.FieldDescriptorProto_TYPE_SFIXED64: true,
+		descriptor.FieldDescriptorProto_TYPE_DOUBLE:   true,
+	},
+
+	// "2	Length-delimited: string, bytes, embedded messages, packed repeated fields"
+	2: {
+		descriptor.FieldDescriptorProto_TYPE_STRING:  true,
+		descriptor.FieldDescriptorProto_TYPE_BYTES:   true,
+		descriptor.FieldDescriptorProto_TYPE_MESSAGE: true,
+	},
+
+	// "3	Start group:	groups (deprecated)"
+	3: {
+		descriptor.FieldDescriptorProto_TYPE_GROUP: true,
+	},
+
+	// "4	End group:	groups (deprecated)"
+	4: {
+		descriptor.FieldDescriptorProto_TYPE_GROUP: true,
+	},
+
+	// "5	32-bit:	fixed32, sfixed32, float"
+	5: {
+		descriptor.FieldDescriptorProto_TYPE_FIXED32:  true,
+		descriptor.FieldDescriptorProto_TYPE_SFIXED32: true,
+		descriptor.FieldDescriptorProto_TYPE_FLOAT:    true,
+	},
+}
+
+// canEncodeType returns true if the wireType is suitable for encoding the descriptor type.
+// See https://developers.google.com/protocol-buffers/docs/encoding#structure.
+func canEncodeType(wireType protowire.Type, descType descriptor.FieldDescriptorProto_Type) bool {
+	if iwt := int(wireType); iwt < 0 || iwt >= len(checks) {
+		return false
+	}
+	return checks[wireType][descType]
+}
+
+// ErrMismatchedWireType describes a mismatch between
+// expected and got wireTypes for a specific tag number.
+type ErrMismatchedWireType struct {
+	Type         string
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+func (mwt *ErrMismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched %q: {TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.Type, mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+func (mwt *ErrMismatchedWireType) Error() string {
+	return mwt.String()
+}
+
+var _ error = (*ErrMismatchedWireType)(nil)
+
+func wireTypeToString(wt protowire.Type) string {
+	switch wt {
+	case 0:
+		return "varint"
+	case 1:
+		return "fixed64"
+	case 2:
+		return "bytes"
+	case 3:
+		return "start_group"
+	case 4:
+		return "end_group"
+	case 5:
+		return "fixed32"
+	default:
+		return fmt.Sprintf("unknown type: %d", wt)
+	}
+}
+
+type UnexpectedField struct {
+	Type     string
+	TagNum   protowire.Number
+	WireType protowire.Type
+}
+
+func (twt *UnexpectedField) String() string {
+	return fmt.Sprintf("UnexpectedField %q: {TagNum: %d, WireType:%q}",
+		twt.Type, twt.TagNum, wireTypeToString(twt.WireType))
+}
+
+func (twt *UnexpectedField) Error() string {
```

Comment:
```suggestion
// String implements the error interface
func (twt *UnexpectedField) String() string {
	return fmt.Sprintf("UnexpectedField %q: {TagNum: %d, WireType:%q}",
		twt.Type, twt.TagNum, wireTypeToString(twt.WireType))
}

// Error implements the error interface
func (twt *UnexpectedField) Error() string {
```

Reply:
String just happens to be used by Error(), but doesn't implement the error interface. Sure, I'll add the comment for .Error() though.